src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -0,0 +1,277 @@
+                                    break;
+                                } else if (previousEndChar == '\n') {
+                                    endOffset--;
+                                    foundNewline = true;
+                                } else if (!foundNewline) {
+                                    endOffset--;
+                                } else {
+                                    break;
+                                }
+                            }
+
+                            // See if this is a selection of complete lines
+                            int startLineNumber = document.getLineNumber(startOffset);
+                            int startLineStartOffset = document.getLineStartOffset(startLineNumber);
+                            if (startLineStartOffset == startOffset) {
+                                int endLineNumber = document.getLineNumber(endOffset);
+                                int endLineEndOffset = document.getLineEndOffset(endLineNumber);
+                                if ((endLineEndOffset == endOffset) || (endLineEndOffset == (endOffset + 1))) {
+                                    // Make sure that it ends with the terminator that was just typed
+                                    String selectionRangeText = StringUtil.trimTrailing(documentChars.subSequence(startOffset, endLineEndOffset).toString());
+                                    return ((startOffset + selectionRangeText.length()) == offset) && selectionRangeText.endsWith(String.valueOf(statementTerminatorChar));
+                                }
+                            }
+                            return false;
+                        }
+                );
+            }
+        }
+
+        // If appropriate, find the enclosing code block to format
+        else if (formattingScope == FormattingScope.CODE_BLOCK) {
+            formatTextRange = LSPCodeBlockProvider.getCodeBlockRange(editor, file, beforeOffset);
+        }
+
+        // If appropriate, use the file text range
+        else if (formattingScope == FormattingScope.FILE) {
+            formatTextRange = file.getTextRange();
+        }
+
+        // If we have a text range now, format it
+        if (formatTextRange != null) {
+            CodeStyleManager.getInstance(project).reformatText(file, Collections.singletonList(formatTextRange));
+            return Result.STOP;
+        }
+
+        return Result.CONTINUE;
+    }
+
+    @NotNull
+    private static Result handleCompletionTriggerTyped(@NotNull Project project,
+                                                       @NotNull Editor editor,
+                                                       @NotNull PsiFile file) {
+        // Just format the completion trigger
+        int offset = editor.getCaretModel().getOffset();
+        // NOTE: Right now all completion triggers are single characters, so this is safe/accurate
+        int beforeOffset = offset - 1;
```

Comment:
For a completion trigger character, we just need to format that single character.